# result - 结果处理

控制器返回结果统一处理，默认返回格式为 JSON

## 依赖模块

- @zenweb/messagecode

## 配置项

| 配置项 | 类型 | 默认值 | 功能 |
| ----- | --- | ----- | ---- |
| failCode | `number` | 无 | 默认失败代码
| failStatus | `number` | 422 | 默认失败HTTP状态码
| json.success | `(ctx: Context, data?: any): any` | `return { data }` | 成功结果包装
| json.fail | `(ctx: Context, err: ResultFail): any` | `return { err.code, err.data, err.message }` | 错误结果包装

## Core 挂载项

无

## Context 挂载项

| 挂载项 | 类型 | 功能 |
| ----- | --- | ---- |
| fail | `(code: string \| number, params?: ParamMap): never` | 失败，输出错误信息并终止代码执行（通常用法）
| fail | `(detail: ResultFailDetail): never` | 失败，输出错误信息并终止代码执行（更多选项）
| success | `(data?: any): any` | 成功，输出结果。（注意：代码会继续执行）

## 可注入对象 

- singleton
  - RenderManager

## 演示
```ts
import { Context, mapping } from 'zenweb';

export class ResultController {
  @mapping()
  hello(ctx: Context) {
    return 'Hello';
  }

  @mapping()
  error(ctx: Context) {
    ctx.fail('error info'); // 在调用 fail 方法后会直接跳出方法并输出
    console.log('这行不会执行');
  }
}
```

### fail 配置

配置 `message-codes.json`

```json
{
  "400": "这是一个错误代码描述",
  "user.username.short: "您的用户名太短：{username}"
}
```

### 失败输出
```ts
ctx.fail(400); // 使用错误代码数值，绝对匹配
ctx.fail('user.username.short', { username: 'AAA' }); // 使用错误代码字符串，递归匹配
// 完全自定义
ctx.fail({
  code: 123,
  message: "自定义",
  status: 200,
});
```
